File: src/main/java/com/elixir/mars/repository/MarsReceiptException.java
package com.elixir.mars.repository;

import java.util.List;
import java.util.Optional;

import com.elixir.mars.repository.models.receipt.MarsError;
import com.elixir.mars.repository.models.receipt.MarsPath;

import lombok.Getter;

public class MarsReceiptException extends RuntimeException {
    @Getter
    private final MarsError error;

    public MarsReceiptException(final String receiptErrorMessage) {
        error = MarsError.builder()
                .message(receiptErrorMessage)
                .build();
    }

    public MarsReceiptException(final Exception exception, final String receiptErrorMessage) {
        this.addSuppressed(exception);
        if (exception instanceof MarsReceiptException) {
            error = ((MarsReceiptException) exception).error;
        } else {
            error = MarsError.builder()
                    .message(receiptErrorMessage)
                    .build();
        }
    }

    public MarsReceiptException(final String receiptErrorMessage, final MarsPath... errorPath) {
        error = MarsError.builder()
                .message(receiptErrorMessage)
                .path(List.of(Optional.ofNullable(errorPath).orElse(new MarsPath[0])))
                .build();
    }

    public MarsReceiptException(
            final Exception exception,
            final String receiptErrorMessage,
            final MarsPath... errorPath) {
        this.addSuppressed(exception);
        if (exception instanceof MarsReceiptException) {
            error = ((MarsReceiptException) exception).error;
        } else {
            error = MarsError.builder()
                    .message(receiptErrorMessage)
                    .path(List.of(Optional.ofNullable(errorPath).orElse(new MarsPath[0])))
                    .build();
        }
    }
}

File: python/mars_repository/mars_receipt_exception.py
from typing import List, Optional

from .models.receipt.mars_error import MarsError
from .models.receipt.mars_path import MarsPath


class MarsReceiptException(RuntimeError):
    def __init__(
        self,
        receipt_error_message: str,
        error_path: Optional[List[MarsPath]] = None,
        exception: Optional[Exception] = None,
    ):
        if exception is not None:
            self.__cause__ = exception

        if isinstance(exception, MarsReceiptException):
            self.error = exception.error
        elif error_path is not None:
            self.error = MarsError(
                message=receipt_error_message,
                type=None,
                path=error_path,
            )
        else:
            self.error = MarsError(
                message=receipt_error_message,
                type=None,
                path=None,
            )

        super().__init__(receipt_error_message)
